pluginutil/pluginutil.go

```diff
@@ -101,28 +156,50 @@ func BuildPluginMap(opt ...Option) (map[string]*PluginInfo, error) {
 // the plugin. In the case of an in-memory plugin it will be nil, however, if
 // the plugin is via RPC it will ensure that it is torn down properly.
 func CreatePlugin(plugin *PluginInfo, opt ...Option) (interface{}, func() error, error) {
+	opts, err := GetOpts(opt...)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing plugin options: %w", err)
+	}
+
+	var file fs.File
+	var name string
+
 	switch {
 	case plugin == nil:
 		return nil, nil, fmt.Errorf("plugin is nil")
 
+	// Prioritize in-memory functions
 	case plugin.InmemCreationFunc != nil:
 		raw, err := plugin.InmemCreationFunc()
 		return raw, nil, err
 
-	case plugin.ContainerFs == nil:
-		return nil, nil, fmt.Errorf("plugin container filesystem is nil")
+	// If not in-memory we need a filename, whether direct on disk or from a container FS
+	case plugin.Path == "":
+		return nil, nil, fmt.Errorf("no inmem creation func and file path not provided")
 
-	case plugin.Filename == "" || plugin.PluginClientCreationFunc == nil:
-		return nil, nil, fmt.Errorf("no inmem creation func and either filename or plugin creation func not provided")
-	}
+	// We need the client creation func to use once we've spun out the plugin
+	case plugin.PluginClientCreationFunc == nil:
+		return nil, nil, fmt.Errorf("plugin creation func not provided")
 
-	opts, err := GetOpts(opt...)
-	if err != nil {
-		return nil, nil, fmt.Errorf("error parsing plugin options: %w", err)
+	// Either we need to have a validated FS to read from or a secure config
+	case plugin.ContainerFs == nil && plugin.SecureConfig == nil:
+		return nil, nil, fmt.Errorf("plugin container filesystem and secure config are both nil")
+
+	// If we have a constructed filesystem, read from there
+	case plugin.ContainerFs != nil:
+		file, err = plugin.ContainerFs.Open(plugin.Path)
+		name = plugin.Path
+
+	// If we have secure config, read from disk
+	case plugin.SecureConfig != nil:
+		file, err = os.Open(plugin.Path)
+		name = filepath.Base(plugin.Path)
+
```

Comment:
Would it be useful to check, perhaps as an option, whether the path provided is a symlink? In Vault's case we want to ensure that the plugin is actually coming from a specific directory. We perform such check in [here](https://github.com/hashicorp/vault/blob/e9adaffead120eb6ccdce76c2b401e7cd9013280/vault/plugin_catalog.go#L537-L543).

Reply:
Maybe?

In Vault there is a configured plugin directory, but additionally you must provide a shasum. It's unclear to me what the value is of disallowing symlinks so long as the sha matches, other than the fact that "plugin directory" is a thing.

This isn't a link to directories, it's to a specific file. If you want to have that be a symlink that can be pointed at different versions of plugins, e.g. for upgrades, I don't think that's an issue so long as the hash matches -- that's the actual security. Even more than that, the checksum is evaluated _after_ the plugin is copied to the new temp location (as bytes, not just a filesystem copy which could possibly copy the symlink). So when we verify the plugin checksum it will definitely not be a symlink.

Note that given the nature of this file-specific construct, if you want to make use of it, nothing prevents you from evaluating symlinks prior to passing the path into this option. So I lean away from building in that functionality centrally given that the real security is from the hash.